### PR TITLE
Grouped Gemm

### PR DESCRIPTION
**Pull request by aazz44ss**
grouped gemm implementation

New type:
hipblasLtExtGroupedGemm_t

New AIP:
hipblasLtExtGroupedGemmCreate()
hipblasLtExtGroupedGemmAlgoGetHeuristic()
hipblasLtExtGroupedGemmMakeArgument()
hipblasLtExtGroupedGemmRun()
hipblasLtExtGroupedGemmDestroy()

./hipBLASLt/build/release/clients/staging/example_hipblaslt_groupedgemm --datatype fp16 --trans_a T --trans_b N -m 1024 -n 1024 -k 1024 -m 512 -n 512 -k 512 -m 2048 -n 1024 -k 512  --sync_count 100 --grouped_gemm 1 --request_solutions 104

### CODE REVIEW

**Comment by jichangjichang** [comment]
I have some questions for MakeArguments() which save the kernel arguments into workspace.
- Do we tell the user how much workspace the kernel needs for the groupgemm problem?
- Do we check the workspace size is enough or not?
- Could you use hipMemcpyAsync() for the kernel arguments copy? hipMemcpy would hurt the perf a lot.

File: library/src/amd_detail/hipblaslt.cpp [hunk 1]
```diff
@@ -353,6 +353,30 @@ catch(...)
...
+    hipblasLtExtGroupedGemmAlgoGetHeuristic(hipblasLtExtGroupedGemm_t        groupedgemm,
+                                            hipblasLtHandle_t                handle,
+                                            hipblasLtMatmulPreference_t      pref,
+                                            int                              requestedAlgoCount,
+                                            hipblasLtMatmulHeuristicResult_t heuristicResultsArray[],
+                                            int*                             returnAlgoCount)
+try
+{
+    if(groupedgemm->gemm_count == 0)
+        return HIPBLAS_STATUS_SUCCESS;
```
[the remark this answers is not in the document]
Comment (aazz44ss) [reply library/src/amd_detail/hipblaslt.cpp]: done

**Comment by aazz44ss** [comment]
> I have some questions for MakeArguments() which save the kernel arguments into workspace.
> 
> * Do we tell the user how much workspace the kernel needs for the groupgemm problem?
> * Do we check the workspace size is enough or not?
No, we haven't do this, will schedule it to next patch.

> * Could you use hipMemcpyAsync() for the kernel arguments copy? hipMemcpy would hurt the perf a lot.
modified makeArgs to use hipMemcpyAsync()

**Comment by jichangjichang** [comment]
> @jichangjichang
> 
> > I have some questions for MakeArguments() which save the kernel arguments into workspace.
> > 
> > * Do we tell the user how much workspace the kernel needs for the groupgemm problem?
> > * Do we check the workspace size is enough or not?
> 
> No, we haven't do this, will schedule it to next patch.
Current APIs seems do not has this feasibility. If we need to provide it in next patch, API may needs to be re-defined.
Therefore, for the APIs definition, it is better to extend this feasibility before the first release.
I would suggest:
HIPBLASLT_EXPORT
hipblasStatus_t hipblasLtExtGroupedGemmCreate(hipblasLtExtGroupedGemm_t*            groupedgemm,
                                              hipblasLtHandle_t                     handle,
                                              std::vector<hipblasLtMatmulDesc_t>&   matmul_descr,
                                              std::vector<float>&                   alpha,
                                              std::vector<void*>&                   A,
                                              std::vector<hipblasLtMatrixLayout_t>& matA,
                                              std::vector<void*>&                   B,
                                              std::vector<hipblasLtMatrixLayout_t>& matB,
                                              std::vector<float>&                   beta,
                                              std::vector<void*>&                   C,
                                              std::vector<hipblasLtMatrixLayout_t>& matC,
                                              std::vector<void*>&                   D,
                                              std::vector<hipblasLtMatrixLayout_t>& matD);


// User can set max workspace size that he can provide in hipblasLtMatmulPreference_t  before call hipblasLtExtGroupedGemmAlgoGetHeuristic().
// After user call hipblasLtExtGroupedGemmAlgoGetHeuristic(), user can check actual workspace size from hipblasLtMatmulHeuristicResult_t.

//Then user can allocate the necessary workspace buffer before call hipblasLtExtGroupedGemmMakeArgument()
HIPBLASLT_EXPORT
hipblasStatus_t hipblasLtExtGroupedGemmMakeArgument(hipblasLtExtGroupedGemm_t    groupedgemm,
                                                    const hipblasLtMatmulAlgo_t* algo,
                                                    void*                                 workspace,
                                                    size_t                                workspaceSizeInBytes,
                                                    hipStream_t                  stream);

**Comment by aazz44ss** [comment]
> > @jichangjichang
> > > I have some questions for MakeArguments() which save the kernel arguments into workspace.
> > > 
> > > * Do we tell the user how much workspace the kernel needs for the groupgemm problem?
> > > * Do we check the workspace size is enough or not?
> > 
> > 
> > No, we haven't do this, will schedule it to next patch.
> 
> Current APIs seems do not has this feasibility. If we need to provide it in next patch, API may needs to be re-defined. Therefore, for the APIs definition, it is better to extend this feasibility before the first release. I would suggest: HIPBLASLT_EXPORT hipblasStatus_t hipblasLtExtGroupedGemmCreate(hipblasLtExtGroupedGemm_t* groupedgemm, hipblasLtHandle_t handle, std::vector<hipblasLtMatmulDesc_t>& matmul_descr, std::vector& alpha, std::vector<void*>& A, std::vector<hipblasLtMatrixLayout_t>& matA, std::vector<void*>& B, std::vector<hipblasLtMatrixLayout_t>& matB, std::vector& beta, std::vector<void*>& C, std::vector<hipblasLtMatrixLayout_t>& matC, std::vector<void*>& D, std::vector<hipblasLtMatrixLayout_t>& matD);
> 
> // User can set max workspace size that he can provide in hipblasLtMatmulPreference_t before call hipblasLtExtGroupedGemmAlgoGetHeuristic(). // After user call hipblasLtExtGroupedGemmAlgoGetHeuristic(), user can check actual workspace size from hipblasLtMatmulHeuristicResult_t.
> 
> //Then user can allocate the necessary workspace buffer before call hipblasLtExtGroupedGemmMakeArgument() HIPBLASLT_EXPORT hipblasStatus_t hipblasLtExtGroupedGemmMakeArgument(hipblasLtExtGroupedGemm_t groupedgemm, const hipblasLtMatmulAlgo_t* algo, void* workspace, size_t workspaceSizeInBytes, hipStream_t stream);

Unlike GSU that workspace is a preference, workspace in grouped gemm is required.
We need workspace information when groupedGemmCreate so we can check if workspace is enough.
Because workspace is required and is hard for users to know workspace usage (workspace varied if number of kernel arguments is different), maybe we should manage it in the backend when groupedGemmCreate and groupedGemmDestroy? so users don't need to worry about it.

**Comment by jichangjichang** [comment]
> Unlike GSU that workspace is a preference, workspace in grouped gemm is required. We need workspace information when groupedGemmCreate so we can check if workspace is enough. Because workspace is required and is hard for users to know workspace usage (workspace varied if number of kernel arguments is different), maybe we should manage it in the backend when groupedGemmCreate and groupedGemmDestroy? so users don't need to worry about it.

If so, we should separate GSU workspace and kernel arguments workspace. Then the GSU workspace is better to be moved to makeArgument() or Run().
Regarding to kernel arguments workspace, is that possible to allocate a large enough size for it in hipblasltHandle?
hipMalloc is blocking call. It is not a good idea to call it everytime in GroupGemmCreate().
If we hide and store it in hipblasltHandle, the kernel arguments workspace can be re-used until handle is destoryed.

I believe that the kernel arguments size should be not huge.  If we pre-define the max the number of GEMM count of hipblaslt GroupGEMM API (ex: 100 or 1000), we can count the required kernel arguments workspace size and pre-allocate it in hipblasltHandle.

**Review by jichangjichang — APPROVED** [review]
LGTM
remember to add hipblaslt CI Test for groupgemm in the coming patch

**Pull request merged**